src/uppy-base/src/plugins/Provider.js

```diff
@@ -7,23 +7,48 @@ const _getName = (id) => {
 }
 
 module.exports = class Provider {
-  constructor (opts) {
+  constructor (core, opts) {
+    this.core = core
     this.opts = opts
     this.provider = opts.provider
     this.id = this.provider
     this.authProvider = opts.authProvider || this.provider
     this.name = this.opts.name || _getName(this.id)
+
+    this.onReceiveResponse = this.onReceiveResponse.bind(this)
+  }
+
+  get hostname () {
+    const uppyServer = this.core.state.uppyServer || {}
+    const host = this.opts.host
+    return uppyServer[host] || host
+  }
+
+  onReceiveResponse (response) {
+    const uppyServer = this.core.state.uppyServer || {}
+    const host = this.opts.host
+    const headers = response.headers
+    // Store the self-identified domain name for the uppy-server we just hit.
+    if (headers.has('i-am') && headers.get('i-am') !== uppyServer[host]) {
+      this.core.setState({
+        uppyServer: Object.assign({}, uppyServer, {
+          [host]: headers.get('i-am')
+        })
+      })
```

Comment:
The `uppyServer` state key is an object of `[original_hostname]: new_hostname` pairs, in case someone does something like

```js
uppy.use(GoogleDrive, { host: 'google.my-uppy-server.com' })
uppy.use(Dropbox, { host: 'dropbox.my-uppy-server.com' })
```

It seems unlikely that anyone would use different uppy-servers for different plugins, but it's _possible_, so this supports it.